src/facilities-context/actions/facilities.tsx

```diff
@@ -0,0 +1,113 @@
+import {
+  deleteFacility,
+  duplicateFacility as duplicate,
+  getFacilities,
+  saveFacility,
+} from "../../database";
+import { Facilities, Facility } from "../../page-multi-facility/types";
+import { FacilitiesDispatch, FacilityMapping } from "../FacilitiesContext";
+
+export const REQUEST_FACILITIES = "REQUEST_FACILITIES";
+export const RECEIVE_FACILITIES = "RECEIVE_FACILITIES";
+export const RECEIVE_FACILITIES_ERROR = "RECEIVE_FACILITIES_ERROR";
+
+export const SELECT_FACILITY = "SELECT_FACILITY";
+export const DESELECT_FACILITY = "DESELECT_FACILITY";
+export const REMOVE_FACILITY = "REMOVE_FACILITY";
+export const CREATE_OR_UPDATE_FACILITY = "CREATE_OR_UPDATE_FACILITY";
+
+export function deselectFacility(dispatch: FacilitiesDispatch) {
+  return () => {
+    dispatch({
+      type: DESELECT_FACILITY,
+    });
+  };
+}
+
+export function selectFacility(dispatch: FacilitiesDispatch) {
+  return (facilityId: Facility["id"]) => {
+    dispatch({
+      type: SELECT_FACILITY,
+      payload: { id: facilityId },
+    });
+  };
+}
+
+export async function fetchFacilities(
+  scenarioId: string,
+  dispatch: FacilitiesDispatch,
+) {
+  dispatch({ type: REQUEST_FACILITIES });
+  try {
+    const facilitiesList: Facilities | null = await getFacilities(scenarioId);
+    const facilities: FacilityMapping = {};
+
+    if (facilitiesList) {
+      facilitiesList.forEach((facility) => {
+        facilities[facility.id] = facility;
+      });
+
+      dispatch({
+        type: RECEIVE_FACILITIES,
+        payload: { ...facilities },
+      });
+    }
+  } catch (error) {
+    // note: should save error message in the state.
+    // currently getFacilities returns null if there is an error
```

Comment:
I may need some guidance on handling the errors / making sure this won't make the app blow up. For now I added `throw error;` at the end of the facility db calls (i.e `saveFacility`, `duplicateFacility`, `getFacilities`, `deleteFacility`), and then I'm using `console.error` in the actions (though this would get printed out twice since we're also doing that in the database functions). 

Do I need to wrap every action call in a `useRejectionToast` so that we show the user the errors?

Reply:
I think you should be able to put your error handling inside the action function and just wrap it around the call to the database function? could set a toast, could capture the error message and put it in state, or whatever is required under the circumstances. 

rejectionToast is really only suitable for failures that don't put the UI into a bad state (e.g. your form submission was rejected) .. I think a failure to load the facility data would exceed that threshold, since the page would be pretty broken without it. It's also something we don't really handle at all currently, because there aren't any expected failure cases, so I don't think you really need to worry about improving that beyond what's necessary to prevent this context from blowing up.